### PR TITLE
Implement bulitin line! macro

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR implements bulitin macro `line!` and add basic infra-structure for other bulitin macros:

1. Extend `MacroDefId` to support builtin macros
2. Add a `quote!` macro for simple quasi quoting.

Note that for support others builtin macros, eager macro expansion have to be supported first, this PR not try to handle it. :)

### CODE REVIEW

File: crates/ra_hir_expand/src/quote.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+            impl Clone for #struct_name {
+                fn clone(&self) -> Self {
+                    Self #list
+                }
+            }
+        };
+
+        assert_eq!(quoted.to_string(), "impl Clone for Foo {fn clone (& self) -> Self {Self {name : self . name . clone () , id : self . id . clone () ,}}}");
+    }
+}
```
Comment (matklad) [inline crates/ra_hir_expand/src/quote.rs]: Neat!

I am not sure if the amount of build-in macros really *requires* us to have quasy-quoting, but it definitely is wort to try!

I'd also would-like to have strongly-typed quasy-quoting for assists some day...

File: crates/ra_hir_expand/src/lib.rs [hunk 2]
```diff
@@ -119,9 +128,22 @@ impl salsa::InternKey for MacroCallId {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct MacroDefId {
+pub enum MacroDefId {
+    DeclarativeMacro(DeclarativeMacro),
+    BuiltinMacro(BuiltinMacro),
+}
```
Comment (matklad) [inline crates/ra_hir_expand/src/lib.rs]:
```rust
struct MacroDefId {
    krate: CrateId,
    ast_id: AstId<ast::MacroCall>,
    kind: MacroDefKind,
}

enum MacroDefKind {
    Declarative,
    BuiltIn(BuiltinExpander),
}
```

File: crates/ra_hir_expand/src/quote.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/ra_hir_expand/src/quote.rs]: If I remember correctly, we would like to expand builtin derive as normal macro expansion. A quasy quoting macro should help there. (Thats why I use `Clone` as a test case )

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**